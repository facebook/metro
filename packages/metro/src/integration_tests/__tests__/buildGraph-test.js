/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+metro_bundler
 * @format
 */

'use strict';

const Metro = require('../../..');
const path = require('path');

jest.unmock('cosmiconfig');

jest.setTimeout(120 * 1000);

it('should build the dependency graph', async () => {
  const entryPoint = path.resolve(
    __dirname,
    '..',
    'basic_bundle',
    'TestBundle.js',
  );

  const config = await Metro.loadConfig({
    config: require.resolve('../metro.config.js'),
  });

  const graph = await Metro.buildGraph(config, {
    entries: [entryPoint],
  });

  expect(
    Array.from(graph.dependencies.entries()).map(([filePath, dep]) => ({
      file: path.basename(filePath),
      types: dep.output.map(output => output.type),
    })),
  ).toEqual([
    {file: 'TestBundle.js', types: ['js/module']},
    {file: 'Bar.js', types: ['js/module']},
    {file: 'Foo.js', types: ['js/module']},
    {file: 'test.png', types: ['js/module/asset']},
    {file: 'AssetRegistry.js', types: ['js/module']},
    {file: 'TypeScript.ts', types: ['js/module']},
  ]);

  expect(graph.dependencies.get(entryPoint)).toEqual(
    expect.objectContaining({
      path: entryPoint,
      inverseDependencies: new Set(),
      output: [
        expect.objectContaining({
          type: 'js/module',
        }),
      ],
    }),
  );

  expect(graph.dependencies.get(entryPoint).output).toMatchSnapshot();
});
