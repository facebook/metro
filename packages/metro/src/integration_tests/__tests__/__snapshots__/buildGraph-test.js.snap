// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build the dependency graph 1`] = `
Array [
  Object {
    "data": Object {
      "code": "__d(function (global, _$$_REQUIRE, _$$_IMPORT_DEFAULT, _$$_IMPORT_ALL, module, exports, _dependencyMap) {
  'use strict';

  var Bar = _$$_REQUIRE(_dependencyMap[0]);

  var Foo = _$$_REQUIRE(_dependencyMap[1]);

  var TypeScript = _$$_REQUIRE(_dependencyMap[2]);

  Object.keys(Object.assign({}, Bar));
  module.exports = {
    Foo: Foo,
    Bar: Bar,
    TypeScript: TypeScript
  };
});",
      "functionMap": Object {
        "mappings": "AAA",
        "names": Array [
          "<global>",
        ],
      },
      "lineCount": 16,
      "map": Array [
        Array [
          2,
          0,
          11,
          0,
        ],
        Array [
          4,
          0,
          13,
          0,
        ],
        Array [
          4,
          6,
          13,
          6,
          "Bar",
        ],
        Array [
          4,
          9,
          13,
          9,
        ],
        Array [
          4,
          12,
          13,
          12,
          "require",
        ],
        Array [
          4,
          23,
          13,
          19,
        ],
        Array [
          4,
          42,
          13,
          0,
        ],
        Array [
          6,
          0,
          14,
          0,
        ],
        Array [
          6,
          6,
          14,
          6,
          "Foo",
        ],
        Array [
          6,
          9,
          14,
          9,
        ],
        Array [
          6,
          12,
          14,
          12,
          "require",
        ],
        Array [
          6,
          23,
          14,
          19,
        ],
        Array [
          6,
          42,
          14,
          0,
        ],
        Array [
          8,
          0,
          16,
          0,
        ],
        Array [
          8,
          6,
          16,
          6,
          "TypeScript",
        ],
        Array [
          8,
          16,
          16,
          16,
        ],
        Array [
          8,
          19,
          16,
          19,
          "require",
        ],
        Array [
          8,
          30,
          16,
          26,
        ],
        Array [
          8,
          49,
          16,
          0,
        ],
        Array [
          10,
          0,
          18,
          0,
          "Object",
        ],
        Array [
          10,
          2,
          18,
          0,
          "Object",
        ],
        Array [
          10,
          8,
          18,
          6,
        ],
        Array [
          10,
          9,
          18,
          7,
          "keys",
        ],
        Array [
          10,
          13,
          18,
          0,
        ],
        Array [
          10,
          32,
          18,
          16,
          "Bar",
        ],
        Array [
          10,
          35,
          18,
          0,
        ],
        Array [
          11,
          0,
          20,
          0,
          "module",
        ],
        Array [
          11,
          2,
          20,
          0,
          "module",
        ],
        Array [
          11,
          8,
          20,
          6,
        ],
        Array [
          11,
          9,
          20,
          7,
          "exports",
        ],
        Array [
          11,
          16,
          20,
          0,
        ],
        Array [
          11,
          19,
          20,
          17,
        ],
        Array [
          12,
          0,
          20,
          18,
          "Foo",
        ],
        Array [
          12,
          4,
          20,
          18,
          "Foo",
        ],
        Array [
          12,
          7,
          20,
          21,
        ],
        Array [
          12,
          9,
          20,
          18,
          "Foo",
        ],
        Array [
          12,
          12,
          20,
          17,
        ],
        Array [
          13,
          0,
          20,
          23,
          "Bar",
        ],
        Array [
          13,
          4,
          20,
          23,
          "Bar",
        ],
        Array [
          13,
          7,
          20,
          26,
        ],
        Array [
          13,
          9,
          20,
          23,
          "Bar",
        ],
        Array [
          13,
          12,
          20,
          17,
        ],
        Array [
          14,
          0,
          20,
          28,
          "TypeScript",
        ],
        Array [
          14,
          4,
          20,
          28,
          "TypeScript",
        ],
        Array [
          14,
          14,
          20,
          38,
        ],
        Array [
          14,
          16,
          20,
          28,
          "TypeScript",
        ],
        Array [
          15,
          0,
          20,
          17,
        ],
        Array [
          15,
          3,
          20,
          0,
        ],
      ],
    },
    "type": "js/module",
  },
]
`;
